cmake: run gen_handles.py before zephyr_prebuilt.elf #39581

tejlmand opened the pull request:
continuation of: #39581 

Main differences to #39581
- now using: `zephyr_pre0.elf`, `zephyr_pre1.elf` as per request, see https://github.com/zephyrproject-rtos/zephyr/pull/39581#pullrequestreview-785489162
- `LINKER_APP_SMEM_UNALIGNED` not applied (and CMake code not invoked) when `CONFIG_USERSPACE=n`
-  CMake linker script generator updated, this should fix the failing twister test cases.

@JordanYates feel free to pick the CMake changes into #39581 and i'll close this afterwards.

Commit:
cmake: update desc of GENERATED_KERNEL_* variables

The previous description of the `GENERATED_KERNEL_*` cmake variables
was incorrect in that these are generated after the second link stage
when `CONFIG_USERSPACE` is enabled.

The sources in these variables are generated from `zephyr_prebuilt.elf`
for inclusion in `zephyr.elf`. However `zephyr_prebuilt.elf` is preceded
by `app_smem_unaligned_prebuilt.elf` when `CONFIG_USERSPACE` is enabled.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
cmake: new global property for generated files

Introduce a new global property for source files that are generated from
the initial Zephyr link (app_smem_unaligned_prebuilt). This source list
is used for files which will introduce address shifts into the final
binary, which need to be present in `zephyr_prebuilt.elf` for
`CONFIG_USERSPACE` scripts to correctly generate `zephyr.elf`.

This resolves #38836.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
cmake: rework linker script generation and linker prebuilt stages

This commit reworks the linker script generation and linking stages in
order to better support fixed section location as required by #38836.

Today we have the following generated linker scripts and the elf output
depending on the system configuration:
- linker_app_smem_unaligned.cmd --> app_smem_unaligned_prebuilt.elf
- linker_zephyr_prebuilt.cmd    --> zephyr_prebuilt.elf
- linker.cmd                    --> zephyr.elf

as not all linker scripts may be created and as there is a need for the
possibility to move gen handles earlier then those stages has been
renamed into more generic names so that with this commit we have:
- linker_zephyr_pre0.cmd --> zephyr_pre0.elf
- linker_zephyr_pre1.cmd --> zephyr_pre1.elf
- linker.cmd             --> zephyr.elf

This also means that is the stage zephyr_pre1 is not needed, then build
can go from `zephyr_pre0.elf` to `zephyr.elf`.

The gen_handles.py has been changed so it now uses `zephyr_pre0.elf` as
input. This ensures that the handles array are final when invoking the
next build and linking stages.

To keep the current behavior of generating the isr table and kobj hash
of what was `zephyr_prebuilt` stage the code blocks contolling isr
generation and kobj hash has been relocated to be located after
app_smem and device handle generation.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>
Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

Commit:
scripts: gen_handles.py: remove size restrictions

With `gen_handles.py` now running on the first pre-built image,
`zephyr_pre0.elf` there is no requirement for the device handle arrays
to remain the same size after processing.

Remove the padding generated in `gen_handles.py`, as well as the
temporary option `CONFIG_DEVICE_HANDLE_PADDING` which was added to work
around this issue.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

ceolin commented:
Nice change using names instead of numbers for the pass phases !

carlescufi commented on CMakeLists.txt:
```diff
@@ -1244,9 +1003,224 @@ if(CONFIG_USERSPACE)
 endif()
 
+if(CONFIG_USERSPACE OR CONFIG_HAS_DTS)
+  list(APPEND LINKER_PASS_${ZEPHYR_CURRENT_LINKER_PASS}_DEFINE "DEVICE_HANDLES_PASS1")
```
Why are you appending `DEVICE_HANDLES_PASS1` even if DTS is disabled?

tejlmand replied:
nice catch.

That would probably have been something I would notice if I was reviewer :laughing: 

Fixed, moved together with gen_handles code where it rightfully belongs.

carlescufi commented on CMakeLists.txt:
```diff
@@ -1244,9 +1003,224 @@ if(CONFIG_USERSPACE)
   )
 endif()
 
+if(CONFIG_USERSPACE OR CONFIG_HAS_DTS)
```
Why isn't `gen_handles.py`  being executed inside this if statement?

tejlmand replied:
because the `gen_handles.py` is only dependent on `HAS_DTS`, not `USERSPACE`.

tejlmand commented on CMakeLists.txt:
```diff
@@ -45,9 +45,12 @@ assert(toolchain_is_ok "The toolchain is unable to build a dummy C file. See CMa
-set(ZEPHYR_PREBUILT_EXECUTABLE zephyr_prebuilt)
+set(ZEPHYR_CURRENT_LINKER_PASS 0)
+set(ZEPHYR_CURRENT_LINKER_CMD  linker_zephyr_pre${ZEPHYR_CURRENT_LINKER_PASS}.cmd)
+set(ZEPHYR_PREBUILT_EXECUTABLE zephyr_pre${ZEPHYR_CURRENT_LINKER_PASS})
```
Done, please comment if you think description can be improved:
https://github.com/zephyrproject-rtos/zephyr/blob/eaf45bc8c8eee3ef2d289cf4adf9aade44e7f697/CMakeLists.txt#L49-L74

tejlmand commented:
Note, regarding this:
https://github.com/zephyrproject-rtos/zephyr/blob/eaf45bc8c8eee3ef2d289cf4adf9aade44e7f697/CMakeLists.txt#L79-L88

I have some plans to take a second round to see if it will be possible to do an overhaul of the linking stages so that it becomes even cleaner what we do when and why.

But I see such work as outside the scope of this PR, so for now I'm not touching the `arch` or `board` code that depends on the `ZEPHYR_PREBUILT_EXECUTABLE` variable.

JordanYates reviewed (commented):
I would prefer to see this as a separate PR that only does the linker conversions, which should make both changes (linker and gen_handles) easier to evaluate.

I feel like there should also be updates to the build system documentation. If its not done when we change it, I feel like its the sort of thing that will be forgotten.

tejlmand commented:
> I would prefer to see this as a separate PR that only does the linker conversions, which should make both changes (linker and gen_handles) easier to evaluate.

You mean like this: https://github.com/zephyrproject-rtos/zephyr/pull/40174 ?

> I feel like there should also be updates to the build system documentation. If its not done when we change it, I feel like its the sort of thing that will be forgotten.

sure, but before updating the text, I would like a review on the approach so that code is approved before updating the text, as to avoid having to update the text if there are comments to the implementation itself.

carlescufi reviewed (approved):
LGTM

JordanYates commented on CMakeLists.txt:
```diff
@@ -1205,9 +1042,222 @@ if(CONFIG_USERSPACE)
+    ${ZEPHYR_INCLUDE_DIRS}
+    )
+
+  add_executable(       ${ZEPHYR_LINK_STAGE_EXECUTABLE} misc/empty_file.c)
```
Minor nit. Whitespace in expr

tejlmand replied:
fixed

JordanYates reviewed (approved):
The final commit comment should be updated, other than that looks good:
```
With `gen_handles.py` now running on `app_smem_unaligned_prebuilt.elf`,
```

JordanYates commented on CMakeLists.txt:
```diff
@@ -1205,9 +1042,222 @@ if(CONFIG_USERSPACE)
+  target_byproducts(TARGET ${ZEPHYR_LINK_STAGE_EXECUTABLE}
+                    BYPRODUCTS ${PROJECT_BINARY_DIR}/${ZEPHYR_LINK_STAGE_EXECUTABLE}.map
+  )
+  set_property(TARGET   ${ZEPHYR_LINK_STAGE_EXECUTABLE} PROPERTY LINK_DEPENDS ${PROJECT_BINARY_DIR}/${ZEPHYR_CURRENT_LINKER_CMD})
```
More whitespace.

tejlmand replied:
fixed

tejlmand commented:
> The final commit comment should be updated, other than that looks good:

Nice catch.
Commit message updated.

As there doesn't seem to be change requests for the implementation itself, I will update the docs accordingly.
`DNM` until docs are updated.

mbolivar-nordic reviewed (approved):
Approach sounds good to me, but I'm trusting @tejlmand with the details. Assuming it works, +1 to remove my DEVICE_HANDLE_PADDING hackarounds.

tejlmand committed:
doc: zephyr build process description

Updated the Zephyr build process description to include description of
all Zephyr link stages.

It extends the guide with description of the intermediate binaries as
well as the post-processing of those binaries.
- Unfixed size binary
- Fixed size binary

Post-processing
- Partition alignment
- ISR table generation

Build stages images has been updated and new images showing the
post-processing stages of the intermediate binaries has been added.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> As there doesn't seem to be change requests for the implementation itself, I will update the docs accordingly.
> `DNM` until docs are updated.

@JordanYates @carlescufi  @ceolin Docs has been updated.

Both with additional text to cover the changes from this PR, as well as updating outdated descriptions.
Several images has been reworked and new images added.

cfriedt commented:
@tejlmand - is this safe to merge now?

tejlmand commented:
> @tejlmand - is this safe to merge now?

I would say so.
@carlescufi has taken a look at the updated docs, and code was already reviewed and approved.

carlescufi reviewed (approved):
Great job!  I particularly appreciate the documentation commit.
